File: app/backend/hawkeye_backend/models/hub.py
"""Hub identity and health. What the iOS Connect screen hits after Bonjour discovery."""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, Field

from hawkeye_backend.models.common import Source, utc_now


class Reachability(StrEnum):
    REACHABLE = "reachable"
    UNREACHABLE = "unreachable"
    DEGRADED = "degraded"
    SIMULATED = "simulated"


class AgentReachability(BaseModel):
    """One agent in the mesh, as seen from the hub.

    The hub reports what it can reach. It does not report ANS verification
    results here: verification is per-claim and lives on the verification stream,
    because an agent that was trusted ninety seconds ago may not be trusted now.
    """

    name: str = Field(description="Agent directory name, e.g. 'agents/people'.")
    ansname: str
    tier: int = Field(ge=1, le=3, description="Build priority tier from agents/CLAUDE.md.")
    reachability: Reachability
    last_seen_at: datetime | None = None
    latency_ms: float | None = None
    detail: str | None = None


class SensorLiveness(BaseModel):
    """Whether CSI is actually flowing.

    `frame_rate_hz` is the field that catches the quiet failure: without a
    traffic generator you get beacons at roughly 10 Hz, which barely resolves
    breathing and never resolves a short motion transient, with every component
    reporting healthy. Below `min_useful_frame_rate_hz` the hub reports degraded.
    """

    source: Source = Field(description="Required. What is behind the sensor interface right now.")
    simulated: bool = Field(description="Derived from source by the hub. Mirrors Provenance.simulated.")
    live: bool = Field(description="Is the sensing pipeline producing frames at all.")
    frame_rate_hz: float | None = None
    min_useful_frame_rate_hz: float = 100.0
    last_frame_at: datetime | None = None
    baseline_healthy: bool
    baseline_age_s: float | None = None
    detail: str | None = None


class CameraStatus(BaseModel):
    """Whether there is a camera, and whether what it last sent is current.

    `linked` and `live` are deliberately two fields rather than one. A link that
    is up while frames have stopped arriving is a real and distinct failure, and
    collapsing it into one boolean would let a stalled camera read as a
    disconnected one, which is a different repair.

    `live` false means the app must not present the last frame as current. That
    is the whole reason this model exists: a frozen picture of an empty room is
    the most dangerous thing this system can put on a screen.
    """

    linked: bool = Field(default=False, description="Is an edge box connected right now.")
    live: bool = Field(
        default=False,
        description="Is the newest frame recent enough to present as current.",
    )
    edge_id: str | None = None
    source: Source | None = Field(
        default=None,
        description=(
            "What the edge claims is producing frames. A video file must not read "
            "as a camera, which is why this is carried rather than assumed."
        ),
    )
    last_frame_age_s: float | None = Field(
        default=None,
        description="Seconds since the newest frame was captured. None if none ever has been.",
    )
    fps: float = Field(default=0.0, description="Frames per second over the last ten seconds.")
    frames_received: int = 0
    frames_dropped: int = Field(
        default=0, description="Inferred from gaps in the edge's frame index."
    )
    detail: str = Field(
        default="No edge camera has connected yet.",
        description="One sentence, written to be printed on a console.",
    )


class HubStatus(BaseModel):
    """GET /v1/hub response.

    The Connect screen verifies the hub with this before entering the main app.
    """

    hub_name: str
    hub_ansname: str = Field(description="The ANSName this hub is anchored to.")
    master_ansname: str
    site_id: str
    site_address: str
    mode: str = Field(description="'simulated' or 'live'. Printed in the app, not hidden.")
    version: str
    healthy: bool = Field(description="True when the app may proceed past the Connect screen.")
    server_time: datetime = Field(default_factory=utc_now)
    uptime_s: float
    sensor: SensorLiveness
    agents: list[AgentReachability]
    active_incident_id: str | None = None
    stream_path: str = "/v1/stream"
    camera: CameraStatus = Field(
        default_factory=CameraStatus,
        description=(
            "The camera's own health, separate from the radio's. The Connect screen "
            "and the live view both read it, and an unreachable camera must render "
            "as unreachable rather than as a still room. Defaulted to the unlinked "
            "state rather than required, because the honest answer before anything "
            "connects is that there is no camera."
        ),
    )

File: app/ios/HawkEye/Models/Hub.swift
import Foundation

/// A Hawk Eye hub discovered on the local network.
///
/// IMPORTANT, and this is a constraint rather than a design choice:
/// **this is not a list of WiFi networks.** iOS does not let a third-party app
/// enumerate nearby SSIDs. Doing that requires the `NEHotspotHelper`
/// entitlement, which Apple grants only to MFi hotspot vendors on request, and
/// which we do not have and will not get. `NEHotspotNetwork` only reports the
/// network you are already joined to, and only with Location permission.
///
/// So the honest mechanism is the one implemented here: the phone is already on
/// the home WiFi, and we browse Bonjour for Hawk Eye hubs advertising
/// `_hawkeye._tcp` on that network. The Connect screen is "which of your hubs",
/// not "which WiFi network".
struct Hub: Sendable, Hashable, Identifiable {

    /// Bonjour service name. Unique on a network, so it works as the identity.
    var id: String
    /// What the resident sees. The hub's own display name.
    var name: String
    /// Resolved endpoint description, for display only.
    var endpoint: String?
    /// 0...1, derived from the interface and RSSI hint the hub publishes in its
    /// TXT record. Nil when the hub does not publish one.
    var signal: Double?
    /// True when this device has paired with this hub before.
    var paired: Bool
    /// The hub's address on this network, published in its TXT record as
    /// `host`.
    ///
    /// **This is carried rather than resolved, deliberately.** The obvious
    /// design is to let `URLSession` resolve the Bonjour service instance, but
    /// `Hub Name._hawkeye._tcp.local.` is a service instance name and not a
    /// hostname, and `URLSession` cannot resolve one. The hub publishes the
    /// address it actually bound and the app dials that. See
    /// `app/backend/hawkeye_backend/discovery.py`.
    ///
    /// Nil for a hub advertising an older TXT record, which cannot be connected
    /// to and is reported as such rather than silently failing to resolve.
    var host: String?
    /// The port the hub serves its API on, from the Bonjour service record.
    /// Nil means fall back to `Config.defaultHubPort`.
    var port: Int?
    /// The hub's ANSName, published in its TXT record. Displayed because it is
    /// the identity everything downstream is anchored to.
    var ansName: String?

    var signalBars: Int {
        guard let signal else { return 0 }
        return max(1, min(4, Int((signal * 4).rounded(.up))))
    }
}

/// `GET /v1/hub`. Hub identity and health, hit after Bonjour discovery.
///
/// `healthy` means "the app may proceed", not "every agent is up". A tier 3
/// agent being unreachable does not block the app; a dead sensing pipeline does,
/// and the app says so rather than drawing an empty house.
struct HubStatus: Codable, Sendable, Hashable {

    /// One agent in the mesh, as seen from the hub.
    ///
    /// **This is reachability, not verification.** Verification is per claim and
    /// lives on the `verification` event, because an agent trusted ninety
    /// seconds ago may not be trusted now.
    struct AgentReachability: Codable, Sendable, Hashable, Identifiable {
        enum Reachability: String, Codable, Sendable, Hashable, CaseIterable {
            case reachable, unreachable, degraded, simulated
        }

        var name: String
        var ansName: String
        /// Build priority tier from `agents/CLAUDE.md`.
        var tier: Int
        var reachability: Reachability
        var lastSeenAt: Date?
        var latencyMs: Double?
        var detail: String?

        var id: String { name }

        enum CodingKeys: String, CodingKey {
            case name, tier, reachability, detail
            case ansName = "ansname"
            case lastSeenAt = "last_seen_at"
            case latencyMs = "latency_ms"
        }
    }

    /// Whether CSI is actually flowing.
    ///
    /// `frameRateHz` against `minUsefulFrameRateHz` is what catches the quiet
    /// failure named in `sensor/CLAUDE.md`: without a traffic generator you get
    /// beacons at roughly 10 Hz, which barely resolves breathing and so cannot
    /// tell a lost signature from a starved capture, with every component
    /// reporting healthy.
    struct SensorLiveness: Codable, Sendable, Hashable {
        var source: Provenance.Source
        /// Derived by the hub from `source`. Mirrors `Provenance.simulated`.
        var simulated: Bool
        var live: Bool
        var frameRateHz: Double?
        var minUsefulFrameRateHz: Double = 100
        var lastFrameAt: Date?
        var baselineHealthy: Bool
        var baselineAgeS: Double?
        var detail: String?

        enum CodingKeys: String, CodingKey {
            case source, simulated, live, detail
            case frameRateHz = "frame_rate_hz"
            case minUsefulFrameRateHz = "min_useful_frame_rate_hz"
            case lastFrameAt = "last_frame_at"
            case baselineHealthy = "baseline_healthy"
            case baselineAgeS = "baseline_age_s"
        }
    }

    var hubName: String
    /// The ANSName this hub is anchored to.
    var hubANSName: String
    var masterANSName: String
    var siteID: String
    var siteAddress: String
    /// `simulated` or `live`. Printed in the app, not hidden.
    var mode: String
    var version: String
    var healthy: Bool
    var serverTime: Date = .distantPast
    var uptimeS: Double = 0
    var sensor: SensorLiveness
    var agents: [AgentReachability] = []
    var activeIncidentID: String?
    var streamPath: String = "/v1/stream"

    enum CodingKeys: String, CodingKey {
        case mode, version, healthy, sensor, agents
        case hubName = "hub_name"
        case hubANSName = "hub_ansname"
        case masterANSName = "master_ansname"
        case siteID = "site_id"
        case siteAddress = "site_address"
        case serverTime = "server_time"
        case uptimeS = "uptime_s"
        case activeIncidentID = "active_incident_id"
        case streamPath = "stream_path"
    }
}

/// Where the Connect screen is in its lifecycle.
enum DiscoveryPhase: Sendable, Hashable {
    /// Browser has not started.
    case idle
    /// Browsing, nothing found yet. The quiet "Looking for your home" state.
    case searching
    /// At least one hub found.
    case found
    /// Verifying the selected hub's identity before entering the app.
    case verifying(hubID: String)
    /// Connected and past the Connect screen.
    case connected(hubID: String)
    /// Browser failed. Local network permission denied is the usual cause.
    case failed(String)
}
